test/functional/p2p_private_broadcast.py

```diff
@@ -0,0 +1,311 @@
+        for i in range(NUM_INITIAL_CONNECTIONS + NUM_PRIVATE_BROADCAST_PER_TX + 2):
+            if i == NUM_INITIAL_CONNECTIONS:
+                # Instruct the SOCKS5 server to redirect the first private
+                # broadcast connection from nodes[0] to nodes[1]
+                self.socks5_server.conf.destinations.append({
+                    "to_addr": "127.0.0.1", # nodes[1] listen address
+                    "to_port": p2p_port(1), # nodes[1] listen port
+                    "node": None,
+                    "requested_to_addr": None,
+                })
+            else:
+                # Create a Python P2P listening node and add it to self.socks5_server.conf.destinations[]
+                listener = P2PInterface()
+                listener.peer_connect_helper(dstaddr="0.0.0.0", dstport=0, net=self.chain, timeout_factor=self.options.timeout_factor)
+                listener.peer_connect_send_version(services=P2P_SERVICES)
+                self.network_thread.listen(
+                    p2p=listener,
+                    # Instruct the SOCKS5 server to redirect a connection to this Python P2P listener.
+                    callback=lambda addr, port: self.socks5_server.conf.destinations.append({
+                        "to_addr": addr,
+                        "to_port": port,
+                        "node": None,
+                        "requested_to_addr": None,
+                    }),
+                    addr="127.0.0.1",
+                    port=ports_base + i)
+                # Wait until the callback has been called and it has done append().
+                self.wait_until(lambda: len(self.socks5_server.conf.destinations) == i + 1)
+
+                self.socks5_server.conf.destinations[i]["node"] = listener
+
+        self.extra_args = [
+            [
+                "-peerbloomfilters", # needed to test replies to MEMPOOL
+                "-cjdnsreachable", # needed to be able to add CJDNS addresses to addrman (otherwise they are unroutable).
+                "-privatebroadcast",
+                f"-proxy={socks5_server_config.addr[0]}:{socks5_server_config.addr[1]}",
+            ],
+            [
+                "-connect=0",
+                f"-bind=127.0.0.1:{p2p_port(1)}=onion", # consider all incoming as coming from Tor
+            ],
+        ]
+        super().setup_nodes()
+
+    def setup_network(self):
+        self.setup_nodes()
+
+    def run_test(self):
+        tx_originator = self.nodes[0]
+
+        observer_inbound = tx_originator.add_p2p_connection(P2PDataStore())
+
+        # Fill tx_originator's addrman.
+        for addr in addresses:
+            res = tx_originator.addpeeraddress(address=addr, port=8333, tried=False)
+            if res["success"]:
+                self.log.debug(f"Added {addr} to tx_originator's addrman")
+            else:
+                self.log.debug(f"Could not add {addr} to tx_originator's addrman (collision?)")
```

Comment:
In 413574e81ef4d608b9b0a2079f81a4e90ee2a341: You can use `-test=addrman` to make addrman deterministic.  Running the test several times I realized that sometimes a collision happens, sometimes not.

Reply:
Done.